configmerge/configmerge.go

```diff
@@ -0,0 +1,244 @@
+	if err != nil {
+		return "", nil, err
+	}
+
+	mergedConfigContent, err := configTree.Merge()
+	if err != nil {
+		return "", nil, err
+	}
+
+	return mergedConfigContent, configTree, nil
+}
+
+func (m *Merger) buildConfigTree(configContent []byte, reference ConfigReference, dir string, depth int, keys []string) (*models.ConfigFileTreeModel, error) {
+	key := reference.Key()
+	keys = append(keys, key)
+
+	if len(configContent) > MaxFileSizeBytes {
+		return nil, fmt.Errorf("max file size (%d bytes) exceeded in file %s", MaxFileSizeBytes, key)
+	}
+
+	if depth > MaxIncludeDepth {
+		return nil, fmt.Errorf("max include depth (%d) exceeded", MaxIncludeDepth)
+	}
+
+	m.filesCount++
+	if m.filesCount > MaxFilesCountTotal {
+		return nil, fmt.Errorf("max include count (%d) exceeded", MaxFilesCountTotal)
+	}
+
+	var config struct {
+		Include []ConfigReference `yaml:"include" json:"include"`
+	}
+	if err := yaml.Unmarshal(configContent, &config); err != nil {
+		return nil, err
+	}
+
+	if len(config.Include) > MaxIncludeCountPerFile {
+		return nil, fmt.Errorf("max include count (%d) exceeded", MaxIncludeCountPerFile)
+	}
+	if m.filesCount+len(config.Include) > MaxFilesCountTotal {
+		return nil, fmt.Errorf("max file count (%d) exceeded", MaxFilesCountTotal)
+	}
+
+	for idx, include := range config.Include {
+		if err := include.Validate(); err != nil {
+			return nil, err
+		}
+		if include.Repository == "" {
+			include.Repository = reference.Repository
+			include.Branch = reference.Branch
+			include.Commit = reference.Commit
+			include.Tag = reference.Tag
+		}
+		config.Include[idx] = include
+
+		if sliceutil.IsStringInSlice(include.Key(), keys) {
+			return nil, fmt.Errorf("circular reference detected: %s -> %s", strings.Join(keys, " -> "), include.Key())
+		}
+	}
+
+	var includedConfigTrees []models.ConfigFileTreeModel
+	for _, include := range config.Include {
+		moduleBytes, err := m.readConfigModule(include, dir, m.repoInfo)
+		if err != nil {
+			return nil, err
+		}
+
+		moduleDir := filepath.Dir(include.Path)
+		moduleConfigTree, err := m.buildConfigTree(moduleBytes, include, moduleDir, depth+1, keys)
+		if err != nil {
+			return nil, err
+		}
+
+		includedConfigTrees = append(includedConfigTrees, *moduleConfigTree)
+	}
+
+	return &models.ConfigFileTreeModel{
+		Path:     key,
+		Contents: string(configContent),
+		Includes: includedConfigTrees,
+		Depth:    depth,
+	}, nil
+}
+
+func (m *Merger) readConfigModule(reference ConfigReference, dir string, repoInfo *RepoInfo) ([]byte, error) {
+	if isLocalReference(reference) {
+		return m.readLocalConfigModule(reference, dir)
+	}
+
+	sameRepo := false
+	if repoInfo != nil {
+		var err error
+		if sameRepo, err = isSameRepoReference(reference, *repoInfo); err != nil {
+			m.logger.Warnf("Failed to check if the reference is from the same repository: %s", err)
+		}
+	}
+
+	if sameRepo {
+		return m.readLocalConfigModule(reference, dir)
+	}
```

Comment:
We discussed that this same repo logic could be removed because if the user enters a repository url then we just simply clone that one. Most probably if the user wants to refer to a file from the same branch/tag/commit then they will just simply enter a path value.